pkg/provider/middleware.go

```diff
@@ -0,0 +1,130 @@
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/containous/maesh/pkg/k8s"
+	"github.com/containous/maesh/pkg/topology"
+	"github.com/containous/traefik/v2/pkg/config/dynamic"
+)
+
+// MiddlewareBuilder is capable of building middlewares for a service.
+type MiddlewareBuilder interface {
+	Build(svc *topology.Service) (*dynamic.Middleware, error)
+}
+
+// AnnotationBasedMiddlewareBuilder build middlewares from service annotations.
+type AnnotationBasedMiddlewareBuilder struct{}
```

Comment:
If the sole purpose of this struct is to implement the interface `MiddlewareBuilder`, I'd rather see this:

```go
type MiddlewareBuilder func(*topology.Service) (*dynamic.Middleware, error) 

func Build(svc *topology.Service) (*dynamic.Middleware, error) {
    // ...
}
```

No need for an interface and an empty struct this way :) And the MiddlewareBuilder type can still be matched by a mock function in the tests.